src/widgets/test-results/loadPatientTestData/helpers.ts

```diff
@@ -0,0 +1,250 @@
+import { getGlobalStore } from "@openmrs/esm-state";
+import {
+  PatientData,
+  ObsRecord,
+  ConceptUuid,
+  ConceptRecord,
+  ObsMetaInfo
+} from "./types";
+
+const retrieveFromIterator = <T>(
+  iteratorOrIterable: IterableIterator<T>,
+  length: number
+): Array<T | undefined> => {
+  const iterator = iteratorOrIterable[Symbol.iterator]();
+  return Array.from({ length }, () => iterator.next().value);
+};
+
+const pageSize = 100;
+
+const cacheStore = getGlobalStore<
+  Record<string, [PatientData, number, string]>
+>("patientResultsDataCache", {});
+
+/**
+ * Adds given user testresults data to a cache
+ *
+ * @param patientUuid
+ * @param data {PatientData}
+ * @param indicator UUID of the newest observation
+ */
+export const addUserDataToCache = (
+  patientUuid,
+  data: PatientData,
+  indicator
+) => {
+  cacheStore.setState({ [patientUuid]: [data, Date.now(), indicator] });
+  const currentStateEntries = Object.entries(cacheStore.getState());
+
+  if (currentStateEntries.length > 3) {
+    currentStateEntries.sort(([, [, dateA]], [, [, dateB]]) => dateB - dateA);
+    cacheStore.setState(Object.fromEntries(currentStateEntries.slice(0, 3)));
+  }
+};
+
+const getLatestObsUuid = async patientUuid => {
+  const request = fhirObservationRequests({
+    patient: patientUuid,
+    category: "laboratory",
+    _sort: "-_date",
+    _summary: "data",
+    _format: "json",
+    _count: "1"
+  });
+
+  return (await request.next().value)?.entry?.[0]?.resource?.id;
+};
+
+/**
+ * Retrieves cached user testresults data
+ * Checks the indicator against the backend while doing so
+ *
+ * @param { string } patientUuid
+ * @param { PatientData } data
+ * @param { string } indicator UUID of the newest observation
+ */
+export const getUserDataFromCache = async (
+  patientUuid: string
+): Promise<PatientData | undefined> => {
+  const [data, , indicator] = cacheStore.getState()[patientUuid] || [];
+
+  if (!!data && (await getLatestObsUuid(patientUuid)) === indicator)
+    return data;
+};
+
+/**
+ * Iterator
+ * @param queries
+ */
+function* fhirObservationRequests(queries: Record<string, string>) {
+  const fhirPathname = "/openmrs/ws/fhir2/R4/Observation";
```

Comment:
I think the leading `/openmrs` is actually a variable. It may be different per configuration. At least that's the purpose of `window.openmrsBase` afaik.